src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs

```diff
@@ -1495,6 +1497,161 @@ static bool extendedTypeIsOrDerivedFrom(NamedTypeSymbol hiddenContainer, NamedTy
+        /// It will remove methods that are "hidden" by any remaining non-methods.
+        /// It will also compact the set (deal with any null entries left in the result).
+        /// </summary>
+        private void MergeExtensionLookupResultsHidingLessSpecificNonMethods(LookupResult current, LookupResult additional, ConsList<TypeSymbol> basesBeingResolved, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+        {
+            if (!current.IsMultiViable || !additional.IsMultiViable)
+            {
+                current.MergePrioritized(additional);
+                return;
+            }
+
+            var currentSymbols = current.Symbols;
+            var additionalSymbols = additional.Symbols;
+
+            var currentCount = currentSymbols.Count;
+            var additionalCount = additionalSymbols.Count;
+            Debug.Assert(currentCount > 0);
+            Debug.Assert(additionalCount > 0);
+
+            for (int i = 0; i < additionalCount; i++)
+            {
+                var additionalSymbol = additionalSymbols[i];
+
+                removeLessSpecificNonMethodCurrentSymbols(additionalSymbol, currentSymbols, out bool isAdditionalSymbolHidden,
+                    currentCount, basesBeingResolved, ref useSiteInfo);
+
+                if (!isAdditionalSymbolHidden)
+                {
+                    currentSymbols.Add(additionalSymbol);
+                }
+            }
+
+            return;
+
+            // Determine whether we want to add the additional symbol
+            // and potentially remove some "hidden" non-method current symbols in the process
+            void removeLessSpecificNonMethodCurrentSymbols(Symbol additionalSymbol, ArrayBuilder<Symbol> currentSymbols, out bool isAdditionalSymbolHidden,
+                int currentCount, ConsList<TypeSymbol> basesBeingResolved, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+            {
+                isAdditionalSymbolHidden = false;
+                for (int j = 0; j < currentCount; j++)
+                {
+                    var currentSymbol = currentSymbols[j];
+                    if (currentSymbol is null)
+                    {
+                        continue;
+                    }
+
+                    if (currentSymbol.ContainingType.ExtendedTypeNoUseSiteDiagnostics is not { } currentContainerUnderlying
+                        || additionalSymbol.ContainingType.ExtendedTypeNoUseSiteDiagnostics is not { } additionalContainerUnderlying)
+                    {
+                        // We wouldn't have gathered these members if we didn't have an extended type for them.
+                        throw ExceptionUtilities.Unreachable();
+                    }
+
+                    if (additionalContainerUnderlying.Equals(currentContainerUnderlying, TypeCompareKind.ConsiderEverything))
```

Comment:
>TypeCompareKind.ConsiderEverything [](http://example.com/codeflow?start=89&length=34)

Do we have a test demonstrating significance of this option?

Reply:
The option has no material effect as type that are loosely equal will not derive from each other and we'll fall through to the last case. We mostly need this comparison to avoid the assertion inside `IsDerivedType`.
We could use a loose comparison here (slightly more efficient) if you prefer. Let me know what you think.

Reply:
> The option has no material effect as type that are loosely equal will not derive from each other and we'll fall through to the last case. We mostly need this comparison to avoid the assertion inside `IsDerivedType`. We could use a loose comparison here (slightly more efficient) if you prefer. Let me know what you think.

I think we should not use options that have "no material effect". If certain differences between the types do not matter, then we should not be paying attention to the differences.